### PR TITLE
runtime-rs: Add basic CH implementation

### PR DESCRIPTION
**Pull request by jodh-intel**
Add a basic runtime-rs `Hypervisor` trait implementation for Cloud Hypervisor (CH).

> **Note:**
>
> This only supports a default Kata configuration for CH currently.

Fixes: #5242.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

### CODE REVIEW

**Comment by jodh-intel** [comment]
## TL;DR

This is a partial implementation only and includes lots of "warts". However,
I'd like to try to get this early code landed to allow others to contribute.

## Intro

This PR introduces a basic runtime-rs CH implementation. It is **not**
ready for general use so I've "hidden" it behind a `cloud-hypervisor`
feature guard for now.

I've decided to take the unusual step (for me, at least) of raising
an "early code" PR since bringing the runtime-rs codebase to feature
parity with the golang runtime for CH is going to take quite a lot
more code. I'd rather land the basics and allow the community to
help iterate on this since it increases visibility, improves
collaboration potential, and avoids endless painful rebases the longer
large PRs remain unmerged.

## Caveats

- Only the default CH Kata Configuration works for now.

- Lots of the values are hard-coded rather than reading them from the
  configuration file (see the gory details in `src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs`).

- There are no tests for the new code yet.

## To enable the Cloud Hypervisor feature

- Modify `src/runtime-rs/crates/runtimes/virt_container/Cargo.toml`

- Enable the `cloud-hypervisor` feature for the `virt_container` crate and its `hypervisor` dependency:
  ```diff
  diff --git a/src/runtime-rs/crates/runtimes/virt_container/Cargo.toml b/src/runtime-rs/crates/runtimes/virt_container/Cargo.toml
  index 2c8226bcd..00bc01ba2 100644
  --- a/src/runtime-rs/crates/runtimes/virt_container/Cargo.toml
  +++ b/src/runtime-rs/crates/runtimes/virt_container/Cargo.toml
  @@ -27,7 +27,7 @@ async-std = "1.12.0"
  
   agent = { path = "../../agent" }
   common = { path = "../common" }
  -hypervisor = { path = "../../hypervisor" }
  +hypervisor = { path = "../../hypervisor", features = ["cloud-hypervisor] }
   kata-sys-util = { path = "../../../../libs/kata-sys-util" }
   kata-types = { path = "../../../../libs/kata-types" }
   logging = { path = "../../../../libs/logging"}
  @@ -36,5 +36,5 @@ persist = { path = "../../persist"}
   resource = { path = "../../resource" }
  
   [features]
  -default = []
  +default = ["cloud-hypervisor"]
   cloud-hypervisor = []
  ```

- Build and install runtime-rs.

/cc @pmores, @amshinde.

**Comment by studychao** [comment]
Hi @jodh-intel , thank you for your wonderful work!!!

I'll take several days to digest the code.

File: src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs [hunk 1]
```diff
@@ -0,0 +1,301 @@
+use anyhow::{anyhow, Context, Result};
```
Comment (Tim-0731-Hzt) [inline src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs:1]: I think maybe it is better to put the `ch_api.rs` into the directory `clh-config`, and rename `clh-config` to `clh-libs`. Because `clh-libs` will be used as a library to manage the VM, which includes config for VM and the function to start or stop VM, attach the device, etc. Just like the `govmm`

File: src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+    PayloadConfig, RngConfig, VmConfig, VsockConfig,
+};
+
+pub async fn cloud_hypervisor_vmm_ping(mut socket: UnixStream) -> Result<()> {
+    task::spawn_blocking(move || -> Result<()> {
+        let response = simple_api_full_command_and_response(&mut socket, "GET", "vmm.ping", None)
+            .map_err(|e| anyhow!(e))?;
+
+        if response.is_some() {
+            warn!(sl!(), "Cloud Hypervisor ping returned: {:?}", response);
```
Comment (Ji-Xinyou) [inline src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs:25]: Does this indicates that the `vmm.ping`'s response is not expected? Otherwise, I would say `info` or `debug` level make more sense.

File: src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+
+pub async fn cloud_hypervisor_vm_create(
+    id: &str,
+    mut socket: UnixStream,
+    shared_fs_devices: Option<Vec<FsConfig>>,
+    pmem_devices: Option<Vec<PmemConfig>>,
+) -> Result<()> {
+    let cfg = cloud_hypervisor_vm_create_cfg(id, shared_fs_devices, pmem_devices).await?;
+
+    //let serialised = serde_json::to_string(&cfg)?;
```
Comment (Ji-Xinyou) [inline src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs:51]: if this is not a hint, please remove this comment line.

File: src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+    let serialised = serde_json::to_string_pretty(&cfg)?;
+
+    task::spawn_blocking(move || -> Result<()> {
+        let data = Some(serialised.as_str());
+
+        let response = simple_api_full_command_and_response(&mut socket, "PUT", "vm.create", data)
+            .map_err(|e| anyhow!(e))?;
+
+        if response.is_some() {
+            warn!(sl!(), "Cloud Hypervisor VM create returned: {:?}", response);
```
Comment (Ji-Xinyou) [inline src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs:61]: ditto

File: src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+    .await?
+}
+
+pub async fn cloud_hypervisor_vm_start(mut socket: UnixStream) -> Result<()> {
+    task::spawn_blocking(move || -> Result<()> {
+        let response = simple_api_full_command_and_response(&mut socket, "PUT", "vm.boot", None)
+            .map_err(|e| anyhow!(e))?;
+
+        if response.is_some() {
+            warn!(sl!(), "Cloud Hypervisor VM boot returned: {:?}", response);
```
Comment (Ji-Xinyou) [inline src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs:75]: ditto

Comment (Ji-Xinyou) [reply src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs:75]: And some dittos below lol

File: src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+        "systemd.log_target=console",
+        "systemd.unit=kata-containers",
+        "systemd.mask=systemd-networkd.service",
+        "systemd.mask=systemd-networkd.socket",
+        "agent.log=debug",
+    ];
+
+    let cmdline = kernel_args.join(" ");
+
+    let kernel = PathBuf::from("/opt/kata/share/kata-containers/vmlinux.container");
```
Comment (Ji-Xinyou) [inline src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs:185]: This should be defined as a `DEFAULT`.

File: src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+        ..Default::default()
+    };
+
+    let serial = ConsoleConfig {
+        mode: ConsoleOutputMode::Tty,
+        ..Default::default()
+    };
+
+    let ip = Ipv4Addr::new(192, 168, 10, 10);
+    let mask = Ipv4Addr::new(255, 255, 255, 0);
```
Comment (Ji-Xinyou) [inline src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs:200]: If this is always used/default in CH, this should be defined as well, such as
```Rust
const DEFAULT_CH_IPV4:  (u8, u8, u8, u8) = (192, 168, 10, 10);
let ip = Ipv4Addr::new(DEFAULT_CH_IPV4.0, DEFAULT_CH_IPV4.1, DEFAULT_CH_IPV4.2, DEFAULT_CH_IPV4.3);
```

Comment (Ji-Xinyou) [reply src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs:200]: And several hard coded values below

File: src/runtime-rs/crates/hypervisor/src/ch/inner.rs [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+    pub(crate) shutdown_tx: Option<Sender<bool>>,
+    pub(crate) shutdown_rx: Option<Receiver<bool>>,
+    pub(crate) tasks: Option<Vec<JoinHandle<Result<()>>>>,
+}
+
+unsafe impl Send for CloudHypervisorInner {}
+unsafe impl Sync for CloudHypervisorInner {}
+
+const CH_DEFAULT_TIMEOUT_SECS: u32 = 10;
```
Comment (Ji-Xinyou) [inline src/runtime-rs/crates/hypervisor/src/ch/inner.rs:59]: please define consts after the `use` section of the file

File: src/runtime-rs/crates/hypervisor/src/ch/inner.rs [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    // Hypervisor runtime directory
+    pub(crate) run_dir: String,
+
+    // Subdirectory of vm_path.
+    pub(crate) jailer_root: String,
+
+    /// List of devices that will be added to the VM once it boots
+    pub(crate) pending_devices: Option<Vec<Device>>,
+
+    pub(crate) _capabilities: Capabilities,
```
Comment (Ji-Xinyou) [inline src/runtime-rs/crates/hypervisor/src/ch/inner.rs:49]: It seems the `_capabilities` is used below, you may change this to `capabilities`

File: src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs (hunk 1 above)
Comment (jodh-intel) [reply src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs]: Agreed. Fixed.

File: src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs (hunk 2 above)
Comment (jodh-intel) [reply src/runtime-rs/crates/hypervisor/src/ch/ch_api.rs]: All these similar comments have been resolved by allowing the caller to handle the response.

**Comment by jodh-intel** [comment]
@Ji-Xinyou - thank you very much for the review. I am very aware of the issues with `ch_api.rs` but, as explained, we'd really like to get this landed to allow others in the community to contribute as soon as possible. It's very painful maintaining large branches like this out of tree, so I am hoping we can agree to land this "as it" now. As the commit details show, this feature is disabled by default (due to the limitations). Once it lands, @amshinde and possibly others such as the [Outreach students](https://github.com/kata-containers/community/wiki/Outreach---Internship-programs) can contribute without having to do constant rebases or the even scarier `git rebase --onto`.

If this is not acceptable to the community, maybe we could create a `staging` branch for large collaborative features like this?

/cc @studychao, @Tim-0731-Hzt, @liubin

**Comment by amshinde** [comment]
@Ji-Xinyou While there is still some work to be done here, as @jodh-intel mentioned, having this landed will help others to start contributing to this work. Launching a cloud hypervisor VM with runtime-rs will be disabled until we get this code in a better shape and other features related to devices and network flushed out. 
If not the main branch, we should consider having this in a separate branch.

wdyt @Ji-Xinyou ?

**Comment by Ji-Xinyou** [comment]
@amshinde @jodh-intel , I think what you guys propose is definitely acceptable, as long as the cloud hypervisor feature is disabled by default until it is more developed.

**Comment by Tim-0731-Hzt** [comment]
> @amshinde @jodh-intel , I think what you guys propose is definitely acceptable, as long as the cloud hypervisor feature is disabled by default until it is more developed.

I agree

File: src/runtime-rs/crates/hypervisor/ch-config/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,18 @@
+[package]
+name = "ch-config"
```
Comment (Tim-0731-Hzt) [inline src/runtime-rs/crates/hypervisor/ch-config/Cargo.toml:2]: I think this package could be named as `ch-libs` because it not only includes config

**Comment by studychao** [comment]
> @Ji-Xinyou - thank you very much for the review. I am very aware of the issues with `ch_api.rs` but, as explained, we'd really like to get this landed to allow others in the community to contribute as soon as possible. It's very painful maintaining large branches like this out of tree, so I am hoping we can agree to land this "as it" now. As the commit details show, this feature is disabled by default (due to the limitations). Once it lands, @amshinde and possibly others such as the [Outreach students](https://github.com/kata-containers/community/wiki/Outreach---Internship-programs) can contribute without having to do constant rebases or the even scarier `git rebase --onto`.
> 
> If this is not acceptable to the community, maybe we could create a `staging` branch for large collaborative features like this?
> 
> /cc @studychao, @Tim-0731-Hzt, @liubin

LGTM! But maybe write somewhere in runtime-rs 's document to tell the community that CLH support is still a WIP:)

**Comment by jodh-intel** [comment]
> GTM! But maybe write somewhere in runtime-rs 's document to tell the community that CLH support is still a WIP:)

@studychao - Good call! I've added a doc commit to address this :smile:

**Review by amshinde — APPROVED** [review]

File: src/runtime-rs/crates/hypervisor/ch-config/src/net_util.rs [hunk 11]
```diff
@@ -0,0 +1,28 @@
+use serde::{Deserialize, Serialize, Serializer};
```
Comment (jiangliu) [inline src/runtime-rs/crates/hypervisor/ch-config/src/net_util.rs:1]: Do we need any copyright header for new files?

File: src/runtime-rs/crates/hypervisor/ch-config/Cargo.toml [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+[dependencies]
+anyhow = "1.0.68"
+serde = { version = "1.0.145", features = ["rc", "derive"] }
+serde_json = "1.0.91"
+tokio = { version = "1.25.0", features = ["sync", "rt"] }
+
+# Cloud Hypervisor public HTTP API functions
+# Note that the version specified is not necessarily the version of CH
+# being used. This version is used to pin the CH config structure
+# which is relatively static.
```
Comment (bergwolf) [inline src/runtime-rs/crates/hypervisor/ch-config/Cargo.toml:17]: Is the CH client API stable? I wonder if it can still work if a different CH version is used.

File: src/runtime-rs/crates/hypervisor/ch-config/src/ch_api.rs [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+            simple_api_full_command_and_response(&mut socket, "PUT", "vm.shutdown", None)
+                .map_err(|e| anyhow!(e))?;
+
+        Ok(response)
+    })
+    .await?
+}
+
+#[allow(dead_code)]
+pub async fn cloud_hypervisor_vm_device_add(mut socket: UnixStream) -> Result<Option<String>> {
```
Comment (bergwolf) [inline src/runtime-rs/crates/hypervisor/ch-config/src/ch_api.rs:87]: Will there be a `device_config` parameter?

File: src/runtime-rs/crates/hypervisor/ch-config/src/net_util.rs (hunk 11 above)
Comment (bergwolf) [reply src/runtime-rs/crates/hypervisor/ch-config/src/net_util.rs:1]: We need a SPDX checker for runtime-rs ;)

Comment (jodh-intel) [reply src/runtime-rs/crates/hypervisor/ch-config/src/net_util.rs:5]: @Ji-Xinyou - Good catch - yes, I've now added the missing headers ;)

File: src/runtime-rs/crates/hypervisor/ch-config/Cargo.toml (hunk 12 above)
Comment (jodh-intel) [reply src/runtime-rs/crates/hypervisor/ch-config/Cargo.toml:21]: My understanding is that it's relatively stable, yes. The pinning is for safety: Although Cloud Hypervisor provides a set of public client API calls, it doesn't provide a public definition of its internal configuration type (`VmConfig`), at least, not in a form we can use [*], so this PR (has to) contain "an implementation" of the _current_ `VmConfig` type. This approach was adopted at the recommendation of the Cloud Hypervisor developers.

So, it _might_ work if we don't pin, but there are no guarantees :cry: 

---

[*] - CH does provide a [YAML format OpenAPI spec file](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/vmm/src/api/openapi/cloud-hypervisor.yaml), but the tooling to convert this into rust is not reliable. Plus, there is some additional complexity around the differences between OpenAPI defined data types defined versus rust lang data types. See https://github.com/cloud-hypervisor/cloud-hypervisor/issues/4732 for further details.

File: src/runtime-rs/crates/hypervisor/ch-config/src/ch_api.rs (hunk 13 above)
Comment (jodh-intel) [reply src/runtime-rs/crates/hypervisor/ch-config/src/ch_api.rs:91]: @bergwolf - Yes. This is just a "nop" implementation until this PR lands. At that point, @amshinde will be helping to actually implement the required functionality.

**Comment by jodh-intel** [comment]
@fengwang666 - Thanks for your comments in the AC meeting. I've now added a comment to the `Cargo.toml` files containing the CH feature that is disabled by default, with a reference to https://github.com/kata-containers/kata-containers/issues/6264:

- https://github.com/kata-containers/kata-containers/pull/6201/files#diff-5f96ba2324977270bd88df163e7b8a9fe2203697889af73e9449143117275818R43-R44
- https://github.com/kata-containers/kata-containers/pull/6201/files#diff-5a4c4930aec7c2c8ea04b03f1df85fd0935e8994f6ffaf29ad2e9d4f047ee797R41-R42

**Comment by jodh-intel** [comment]
:green_circle: :green_circle: :green_circle: :green_circle: :green_circle: :green_circle: :green_circle: :green_circle: :green_circle: :green_circle:
:green_circle: CI is all green folks :smile: :green_circle: 
:green_circle: :green_circle: :green_circle: :green_circle: :green_circle: :green_circle: :green_circle: :green_circle: :green_circle: :green_circle:


We just need one more ack so we can land this to make contributing to the new Cloud Hypervisor feature easier.

**Comment by ariel-adam** [comment]
Awesome :-)

**Comment by fidencio** [comment]
Please, wait till the release is over and then go ahead to merge it.

**Comment by bergwolf** [comment]
Release is out. Let's merge this :)

**Comment by jodh-intel** [comment]
Thanks @bergwolf ! :smile:
